csi/moac/volume.ts

```diff
@@ -247,72 +234,37 @@ export class Volume {
         this._setState(VolumeState.Faulted);
         return;
       }
-
-      if (!this.publishedOn) {
-        // If the volume hasn't been published we can't do anything more than what
-        // we have done (that is maintain required # of replicas). When we create
-        // the nexus, it may find out that some of the replicas are unusable, but
-        // we don't know that now.
-        this._setState(VolumeState.Healthy);
-        return;
-      }
     }
 
     // check that replicas are shared in the way they should be
-    let localNode: string = (this.nexus) ? this.nexus.node.name : this.publishedOn;
-    for (const nodeName in this.replicas) {
-      const replica = this.replicas[nodeName];
-      if (replica.isOffline()) {
-        continue;
-      }
-      let share;
-      const isLocal = replica.pool!.node.name === localNode;
-      if (isLocal && replica.share !== 'REPLICA_NONE') {
-        // make sure that replica that is local to the nexus is accessed locally
-        share = 'REPLICA_NONE';
-      } else if (!isLocal && replica.share === 'REPLICA_NONE') {
-        // make sure that replica that is remote to nexus can be accessed
-        share = 'REPLICA_NVMF';
-      }
-      if (share) {
-        try {
-          await replica.setShare(share);
-          delete this.nodeBlackList[nodeName];
-          // fsa will get called again because the replica was modified
-          return;
-        } catch (err) {
-          this.nodeBlackList[nodeName] = true;
-          log.error(
-            `Failed to set share protocol to ${share} for replica "${replica}": ${err}`
-          );
-        }
-      }
+    let nexusNode, replicaSet;
+    try {
+      [nexusNode, replicaSet] = await this._ensureReplicaShareProtocols();
+    } catch (err) {
+      log.warn(err.toString());
+      return;
     }
 
-    // If we don't have a nexus and the volume is published, then try to create one
+    // If we don't have a nexus and we should have one then create it
     if (!this.nexus) {
-      assert(this.publishedOn);
-      let nexusNode = this.registry.getNode(this.publishedOn);
-      if (nexusNode && nexusNode.isSynced()) {
-        let replicas = [];
-        for (let nodeName in this.replicas) {
-          if (!this.replicas[nodeName].isOffline() && !this.nodeBlackList[nodeName]) {
-            replicas.push(this.replicas[nodeName]);
+      if (
+        this.publishedOn ||
+        replicaSet.length !== this.replicaCount
```

Comment:
If the user reduces the number of replicas, we don't need to actually bring up a nexus?

Reply:
It's true that it's not strictly necessary. It's more kind of a double check that we don't do anything stupid. The case is as follows. All that moac sees is that there are n replicas and there should be n - 1. It does not know if all replicas are sound and healthy. They should be but perhaps not in all cases. Then creating a nexus to tell us what the nexus thinks about those replicas (i.e. is replica reachable from nexus) before we try to destroy any of them, can be seen as another safety check. It cannot tell us if any replica is out of sync because we haven't implemented generation numbers still, but that would be also possible in the future. But I admit that I'm being maybe too much paranoid. So I will defer to reviewers opinion on this. Although that the code is simpler as it is now, adding a new block of code for removing replica when there is no nexus would not be probably too much work to do.

Reply:
Yeah I suppose it doesn't hurt to do it, and we can check if the replicas are still "alive". Although if any replica is not "alive" then the nexus creation will fail IIRC, but that's a separate issue altogether.

Reply:
yea, that's true and I did not think about it. It's like stepping through a mine field now  :-D  So we might end up in a situation when we simply cannot remove a replica from volume, because the replica that appears to be ok is not reachable from nexus, that's why nexus creation fails, that's why we can't remove any replica 🤔  not a nice case. On the other hand, moac is not the only piece of puzzle in this game and it is a general problem that can bite us at many places (i.e. also when we try to scale up the volume).

Reply:
Yeah, so if we fix that issue, then bring up the nexus to reduce a replica should not pose any issue. Meanwhile if we wish to decrease the replicas when 1 replica is not "alive", then we could always manually remove it from the list of children?

Reply:
I have never tried that (editing replicas in msv). In general it is the part of crd that users shouldn't touch. Though IMHO it is probably acceptable to live with this corner case until nexus create is fixed.